src/map/npc.cpp

```diff
@@ -945,122 +972,128 @@ int npc_touch_areanpc_sub(struct block_list *bl, va_list ap)
  *------------------------------------------*/
 int npc_touchnext_areanpc(struct map_session_data* sd, bool leavemap)
 {
-	struct npc_data *nd = map_id2nd(sd->touching_id);
+	struct npc_data *nd;
+	int k, found = 0;
 	short xs, ys;
 
-	if( !nd || nd->touching_id != sd->bl.id )
+	if (sd->npc_ontouch_.empty())
 		return 1;
 
-	xs = nd->u.scr.xs;
-	ys = nd->u.scr.ys;
+	for ( k = 0; k < sd->npc_ontouch_.size(); k++ ) {
```

Comment:
instead of iterating through the list multiple time (once for the actual function, second time to delete the id with `std::erase`). You should use `std::remove_if` like 
````cpp
std::remove_if(sd->npc_ontouch_.begin(), sd->npc_ontouch_.end(), []() {
    nd = map_id2nd(sd->npc_ontouch_[k]);
    if (!nd) {
        return true;
    } else {
        xs = nd->u.scr.xs;
        ys = nd->u.scr.ys;

        // note : hiding doesn't reset the previous trigger status
        // player must leave the area to reset nd->touching_id on official
        if (sd->bl.m != nd->bl.m ||
            sd->bl.x < nd->bl.x - xs || sd->bl.x > nd->bl.x + xs ||
            sd->bl.y < nd->bl.y - ys || sd->bl.y > nd->bl.y + ys ||
            leavemap)
        {
            char name[EVENT_NAME_LENGTH];

            if (nd->touching_id && nd->touching_id == sd->bl.id) {// empty when reload script
                found = 1;
                nd->touching_id = 0;
                safesnprintf(name, ARRAYLENGTH(name), "%s::%s", nd->exname, script_config.ontouch_event_name);
                map_forcountinarea(npc_touch_areanpc_sub,nd->bl.m,nd->bl.x - xs,nd->bl.y - ys,nd->bl.x + xs,nd->bl.y + ys,1,BL_PC,sd->bl.id,name);
            }


            return true;

        }
    }
    
	return false;
});
````